### DISCUSSION TITLE
Some corrections for MRAF in the package

### DISCUSSION THREAD

**Discussion by Shuyun1247** [slmsuite/slmsuite#64] [Ideas]
Hello,

I've been using your package, and it's been working great for me. However, I encountered some issues when using the MRAF method in your code, specifically when setting the signal region be a rectangular "null_region". I've identified some areas in `algorithm.py` that, upon modification, resolve these issues. Here are the changes I suggest:

1. **Modification in `ijcam_to_knmslm` function**:
   In `algorithms.py`, find the definition of the function `ijcam_to_knmslm`. Before the affine transformation, add the line `b = b.reshape(-1)` to ensure the correct shape for `b`.
   
   **Original**:
   ```python
   cp_img = cp.array(img, dtype=self.dtype)
   cp.abs(cp_img, out=cp_img)
   
   # Perform affine.
   ...
   ```
   **Modified**:
   ```python
   cp_img = cp.array(img, dtype=self.dtype)
   cp.abs(cp_img, out=cp_img)

   b = b.reshape(-1)

   # Perform affine.
   ...
   ```
2. **Modification related to `null_knm` and `null_region_knm`**:
   Find the following section in `algorithms.py` and modify it as follows to account for both `null_knm` and `null_region_knm`.

**Original**:
   ```python
   if self.null_knm is None:
    ...
   ```
**Modified**:
   ```python
   if self.null_knm is None and self.null_region_knm is None:
    ...
   ...
   ```
...

I'm not entirely sure if I was using the `SpotHologram` incorrectly, but after making these corrections, the method worked properly for both "null_vector" and "null_region" for us. I hope these suggestions are helpful. Let me know if you need any more details or if there are any concerns.

Thank you!

**Comment by ichristen** [comment]
Hi @Shuyun1247,

Thank you for your interest in slmsuite!

Could you provide a test case that results in a bug? Your fix to 1 makes me suspicious that there's some versioning or package inconsistency with numpy/cupy transformations (or it could just be a bug). In similar cases, I've added failsafes so the code works on both versions. Your fix to 2 looks reasonable.

This is definitely a bug/issue, so I'm moving this discussion to Issues. The addition of 3D spots #41 in the next release is going to mess with a lot of the internals of `SpotHologram`, so I will resolve this issue as part of that pull request.

Thanks for this feedback! Let us know if you find anything else similar to this (every bug reported is a bug squashed), or if there are any features of particular interest.

Best,
Ian

**Comment by ichristen** [comment]
See #65
